module/support/logging/MessageLogHandler/src/MessageLogHandler.cpp

```diff
@@ -0,0 +1,91 @@
+            out.identifiers.dsl      = in.identifiers.dsl;
+            out.identifiers.function = in.identifiers.function;
+
+            out.reaction_id       = uint64_t(in.reaction_id);
+            out.task_id           = uint64_t(in.task_id);
+            out.cause_reaction_id = uint64_t(in.cause_reaction_id);
+            out.cause_task_id     = uint64_t(in.cause_task_id);
+            out.emitted           = in.emitted;
+            out.started           = in.started;
+            out.finished          = in.finished;
+
+            if (in.exception) {
+                try {
+                    std::rethrow_exception(in.exception);
+                }
+                catch (const std::exception& ex) {
+                    out.exception = ex.what();
+                }
+                catch (...) {
+                    // We don't actually want to crash
+                }
+            }
+
+            return out;
+        }
+
+        message::nuclear::LogLevel to_message(const NUClear::LogLevel& level) {
+            switch (level) {
+                case NUClear::LogLevel::TRACE: return message::nuclear::LogLevel::TRACE;
+                case NUClear::LogLevel::DEBUG: return message::nuclear::LogLevel::DEBUG;
+                case NUClear::LogLevel::INFO: return message::nuclear::LogLevel::INFO;
+                case NUClear::LogLevel::WARN: return message::nuclear::LogLevel::WARN;
+                case NUClear::LogLevel::ERROR: return message::nuclear::LogLevel::ERROR;
+                case NUClear::LogLevel::FATAL: return message::nuclear::LogLevel::FATAL;
+                case NUClear::LogLevel::UNKNOWN:
+                default: return message::nuclear::LogLevel::UNKNOWN;
+            }
+        }
+
+    }  // namespace
+
+    MessageLogHandler::MessageLogHandler(std::unique_ptr<NUClear::Environment> environment)
+        : Reactor(std::move(environment)), config{} {
+
+        on<Configuration>("MessageLogHandler.yaml").then([this](const Configuration& cfg) {
+            // Use configuration here from file MessageLogHandler.yaml
+            this->log_level = cfg["log_level"].as<NUClear::LogLevel>();
+        });
+
+        on<Trigger<NUClear::message::LogMessage>>().then([this](const NUClear::message::LogMessage& log_msg) {
+            // Skip if we are not logging this level
+            if (log_msg.level < config.min_level || (!config.log_all && log_msg.level < log_msg.display_level)) {
+                return;
+            }
+
+            auto msg           = std::make_unique<message::nuclear::LogMessage>();
+            msg->timestamp     = NUClear::clock::now();
+            msg->level         = to_message(log_msg.level);
+            msg->display_level = to_message(log_msg.display_level);
+            if (log_msg.task) {
+                msg->reaction_statistics = to_message(*log_msg.task);
+            }
+            emit(msg);
+        });
+
+        on<Trigger<NUClear::message::ReactionStatistics>>().then(
```

Comment:
Why do we need to pass the message into the `to_message` function here?

Do the NUClear library messages not get included in the NUsight protobut stuff?

Reply:
I think the current setup for NUbots and NUsight doesn't work with messages outside the `shared/message` directory.

@TrentHouliston or @Bidski can say definitively.

Reply:
The protobuf stuff is separate from the c++ NUClear stuff.
I believe the enums are probably the same value and you could just directly cast it rather than going through this process.
The advantage of this process is any mismatch is a compile error, not a future being confused why things are different error.